### ISSUE TITLE
Output formats and what produces the text output format?

### ISSUE THREAD

**Issue by karlmdavis** [HdrHistogram/HdrHistogram#170]
Apologies for the dumb question, but I'm wondering if someone might explain the various Histogram output formats to me: what are the different output formats and what (Java) code produces them?

The histogram format I'm most familiar is this one, which I call the "text" format: <https://github.com/HdrHistogram/HdrHistogram/blob/master/GoogleChartsExample/example1.txt>. But I see documentation that also mentions the `.hgrm` format. What part of the code produces that output? (The "documentation" I'm referring to is mostly the note at the bottom of this page: <http://hdrhistogram.github.io/HdrHistogram/plotFiles.

I ask because I'd like to use the Rust port of the library to produce files that can be plotted using that handy-dandy webpage, but I can't tell if the Rust code already exports/serializes to one of the supported formats or if I'll have to port that feature first, myself.

Any help/guidance would be much appreciated.

**Comment by giltene** [comment]
The .hgrm format is used for a log of interval histograms, it is a text file with one text line per time interval, with each interval histogram represented in a deflated serialized form. Most HdrHistogram language implementations (Java, C, C#, JS, and Python, and Rust) support the V2+Deflate interval log format used in .hgrm logs.

This logging format is common in monitoring and measurement. You can find some examples of .hgrm interval logs in e.g./ the [test resources for the java implementation](https://github.com/HdrHistogram/HdrHistogram/blob/master/src/test/resources/org/HdrHistogram/jHiccup-2.0.7S.logV2.hlog). Tools that can parse these logs and display them include e.g. [HistogramLogAnalyzer](https://github.com/HdrHistogram/HistogramLogAnalyzer).

For Rust specifically, you can find documentation about the interval logs and the serialized histogram format 
https://docs.rs/hdrhistogram/7.1.0/hdrhistogram/serialization/interval_log/index.html
https://docs.rs/hdrhistogram/7.1.0/hdrhistogram/serialization/index.html

While some language implementations also support outputting a single histogram in the percentile distribution text format (e.g. what the java [outputPercentileDistribution](https://www.javadoc.io/static/org.hdrhistogram/HdrHistogram/2.1.12/org/HdrHistogram/AbstractHistogram.html#outputPercentileDistribution-java.io.PrintStream-java.lang.Double-) does), which is what you call the "text" format above, I believe that the Rust implementation does not directly implement this right now. However, there it appears to include a [code example](https://github.com/HdrHistogram/HdrHistogram_rust/blob/master/examples/cli.rs#L139) that does output that format. 

Alternatively, you can use the Rus port's interval log format to output a single interval histogram, and then use the Java library's handy [HistogramLogProcessor](https://github.com/HdrHistogram/HdrHistogram/blob/master/HistogramLogProcessor) tool included with the Java version to post-process that histogram log and convert it to the percentile distribution form (see the [JavaDoc for the HistogramLogProcessor class](https://www.javadoc.io/static/org.hdrhistogram/HdrHistogram/2.1.12/org/HdrHistogram/HistogramLogProcessor.html) for a description of functionality and command-line options).

HTH.

**Comment by darach** [comment]
We had a similar need for plotFiles compatible output on tremor for micro
benchmark output so we wrote a function to achieve this:

https://github.com/wayfair-tremor/tremor-runtime/blob/main/src/offramp/blackhole.rs#L136

HTH

**Comment by marshallpierce** [comment]
+1 on all of that, with a bit more about the Rust example CLI: it produces _almost_ the same output as `outputPercentileDistribution. When I wrote that I had just spent far too much time tracking down some floating point artifacts that a sharp-eyed user had pointed out (which turned out to be inherent in floating point, with not much we can do about them), so it outputs _quantiles_ not _percentiles_ because in certain cases the extra multiplication would introduce minor artifacts. That is, it would output `0.92` instead of `92`.

If you really want a percentile output, it would be easy to add as an additional option in the CLI, but I think .hgrm (interval log) is really the way to go. It's compact, standardized, supported by the plotting tool, and you can easily write glue in either Rust or Java to turn it into quantile or percentile text output for human inspection if you want.

**Comment by ahothan** [comment]
All valid questions by the OP.
@giltene note that there is a discrepancy of extension naming, I was initially confused by the ".hgrm" extension (never heard about it until I see it now at the bottom of this plotFile web page) but I know about the example that is in the Java repo which is ".hlog". We should make it consistent and use only 1 extension (preferably .hlog).
We may also need a formal name for the base64 compressed string that represents 1 histogram (I'm using this compressed format in my json results and did not know how exactly to call that format).

We really need this more formal spec/definition ;-)

**Comment by giltene** [comment]
@ahothan i was wrong about meaning of the .hgrm extension name convention (late night posting), and will edit original reply to avoid confusing future readers. .hlog is the extension name used by convention for interval histogram logs, the .hgrm extension is used for textual percentile distribution of a single histogram, in a format easily consumable by tools and plotters (importing to excel, gnuplot, etc.) Two reference tools I have out there that produce files using these conventions are [jHiccup](https://github.com/giltene/jHiccup) (produces .hlog output With .hlog in the default log names) and [HistogramLogProcessor](https://www.javadoc.io/doc/org.hdrhistogram/HdrHistogram/latest/index.html) or jHiccupLogProcessor (see jHiccup README) (produces log files with  .hgrm extension name along with other output).

+1 on coming up with a formal name for the base64 encoded compressed serialized histogram thing (the V2+Deflate format). How about base64Histogram?

And +1000 on the spec/definition. Some day we’ll be stuck at home long enough to do this. Oh, wait...

**Comment by karlmdavis** [comment]
> +1 on coming up with a formal name for the base64 encoded compressed serialized histogram thing (the V2+Deflate format). How about base64Histogram?

What about `".hgrm64"`?

**Comment by karlmdavis** [comment]
> If you really want a percentile output, it would be easy to add as an additional option in the CLI, but I think .hgrm (interval log) is really the way to go. It's compact, standardized, supported by the plotting tool, and you can easily write glue in either Rust or Java to turn it into quantile or percentile text output for human inspection if you want.

@marshallpierce (or whomever): A couple of questions on this:

1. I assume you meant the "`.hlog`" format, instead?
2. Which plotting tool supports the "`.hlog`" format? Based on its documentation (at the bottom of the page), I'm guessing that <http://hdrhistogram.github.io/HdrHistogram/plotFiles.html> doesn't, right?

**Comment by karlmdavis** [comment]
> The hgrm format (which is used by convention in .hgrm files) is the a textual percentile distribution of a single histogram. It is designed to be easily consumable not only by plotting tools that like plotFiles, but also by common plotters like excel and gnuplot. Specifically, its additional 1/(1-Percentile) column reports useful values that require no further mathematical conversion to be plotted a percentile axis to produce the typical visual representation we use in percentile charts. Code that produces this format can be found in e.g. the java implementation of HdrHistogram (and exists in several other language implementations, including C, C#, python, and js).

@giltene (or whomever): Am I correct in assuming that conversion to the "`.hgrm`" format is a one-way trip; that there's no way to convert that file format back into an actual deserialized `Histogram` object/struct?

**Comment by marshallpierce** [comment]
@karlmdavis ah, you are correct -- I had thought that the web tool could plot interval logs or at least an individual serialized histogram, but it looks like it only supports the percentile text format. I think the TS implementation supports the serialized format these days, so perhaps that's a relatively straightforward improvement.

**Comment by karlmdavis** [comment]
>> +1 on coming up with a formal name for the base64 encoded compressed serialized histogram thing (the V2+Deflate format). How about base64Histogram?
>
> What about ".hgrm64"?

Thinking about it some more (as I'm trying to document my code and what it's _actually_ doing), I'd propose being explicit with these names and extensions throughout the documentation and examples:
* "Serialized Histogram": `.hgrm64`
* "Serialized Compressed Histogram": `.hgrm64z`

And I'd propose using the following names for the other formats consistently through the docs:
* "Histogram Percentiles Text Export": `.hgrm`
* "Histogram Interval Log": `.hlog`

Those are all the formats supported across the base library and its ports, right?

I try real hard to not be particularly clever about this stuff: straightforward is usually better.

**Comment by karlmdavis** [comment]
Darn, I guess it gets even more complicated when you contemplate that the code also supports producing a non-Base64'd serialized histogram, too. I'm actually not even sure which is the default: Base64'd or not.

Sigh.

**Comment by marshallpierce** [comment]
A serialized histogram sans base64 is a reasonable standalone thing, e.g. to ship over the wire to some diagnostic tool. I would guess that the base64 flavor is primarily (only?) used in the interval log format.

**Comment by giltene** [comment]
> @giltene (or whomever): Am I correct in assuming that conversion to the "`.hgrm`" format is a one-way trip; that there's no way to convert that file format back into an actual deserialized `Histogram` object/struct?

Yes. For two reasons.

.hlog files contain multiple histograms. One per time interval. .hgrm files show a textual representation of percentile distribution of a single histogram (the histogram that us the sum of the histograms in the range of time processed from the .hlog file, which by default is the whole file).

The individual histograms logged in .hlog files are complete HdrHistograms. They have all the information that an HdrHistogram has, maintain the same precision level as the histogram That was logged, and can be used to losslessly reconstruct that logged hdrhistogram. The percentile distribution output contains much less information, it summarizes the percentile distributions in the logged histogram, but any histogram reconstructed from that information would very likely differ from the histogram being logged ( it would have the same percentile distribution at the reported percentile levels in the .hgrm text, but that’s it).

Think of it this way:

.hgrm is a summary of information in a histogram. It is human readable, and can be easily invested into e.g. spreadsheets or generic plotting tools.

.hlog (and the individual V2+deflate base64 lines within it) contains actual, complete hdrhistograms. It can only be invested by things that can reconstruct hdrhistograms (and humans and excel sheets don’t do that very well).

(This distinction is somewhat complicated by the fact that hdrhistograms are themselves “approximations” of the data that was accumulated into them, but that approximation follows the rules specified in the creation of the hdrhistogram (the precision selected), and loses nothing beyond that).

**Comment by ahothan** [comment]
@marshallpierce you're right that the binary version (not b64 encoded) could be useful on the wire, although in my own use I just ended up including the b64 encoded version inside a json dictionary along with other data, I think the b64 encoded version is a lot more convenient to handle in practice (nesting in json, csv, hlog...)

The compressed format (b64 or not) is the most precise format and it might be a bit misleading to use the same extension start than the .hgrm prefix which is less precise (see below).
I'd suggest to use ".hdrh" for the compressed base 64 encoded histogram. In terms of terminology, "hdrh format" would refer to the base64 encoded version by default, while "binary hdrh format" would refer to the binary version which is not base64 encoded. These 2 formats are reversible without loss:
binary hdrh ----base64 encoding---> hdrh
hdrh ---base64 decoding---> binary hdrh
If we really need an extension for the binary version, ".hdrhz"
We could explore other options that are more logical but they tend to yield longer extensions (.hdrhz, .hdrz64 ...)
 
The compression algorithm used (pre base 64 encoding) is pretty custom-taylored for this project and that is why it is hard to put an extension to it.
hlog is just a sequence of hdrh along with some extra metadata (timestamp...)

**Comment by ahothan** [comment]
To add to @giltene , .hgrm is generally more than sufficient for plotting. However, the compressed format is best for any other more advanced use such as lossless histogram aggregation which is really important for applications that need discrete vs aggregated latency histograms and rolling latency stats (e.g. histogram for last 5m when you get 1m interval).

**Comment by giltene** [comment]
I don’t think we need another file name (file extension) conversion. We need a term that is a a short way to say “HdrHistogram V2+Deflate compressed base64 encoded serialized form” (the last item on the individual lines of a .hlog file, the thing you would probably put in json, csv, etc.). We probably also want a term that is a short way to say “HdrHistogram V2+Deflate compressed serialized form” (the non-base64 thing, the thing you would likely use in BLOBs and in binary wire forms).

“Serialized Histogram” and “Binary Serialized Histogram” May be good starts. Or maybe “Serialized HdrHistogram” and “Binary Serialized HdrHistogram”. But a newly invented one word shorthand thing would be cool to find.
HistoBlob? 
[Which can stand for both the binary firm and the base64 encoded version, with “base64 HistoBlob” used if/when clarification is needed (E.g. in .explaining .hlog,  but no clarification needed for e.g. json, because binary blobs in json are typically base64 encoded anyway).]

**Comment by ahothan** [comment]
I agree we don't really need an extension as in all practical cases that blob is nested inside another file that has an extension.
histoblob is good.
by default it will refer to the base 64 version which is the most common format.
if needed, the binary version of it can be called "binary histoblob".

**Comment by karlmdavis** [comment]
"histoblob" :laughing: 

:+1:
